src/Java.Interop.Tools.JavaTypeSystem/Adapters/JavaXmlApiImporter.cs

```diff
@@ -0,0 +1,370 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+using Java.Interop.Tools.JavaTypeSystem.Models;
+
+namespace Java.Interop.Tools.JavaTypeSystem
+{
+	public class JavaXmlApiImporter
+	{
+		public static JavaTypeCollection ParseString (string xml, JavaTypeCollection? collection = null)
+		{
+			var doc = XDocument.Parse (xml, LoadOptions.SetBaseUri | LoadOptions.SetLineInfo);
+
+			return Parse (doc, collection);
+		}
+
+		public static JavaTypeCollection Parse (TextReader reader, JavaTypeCollection? collection = null)
+		{
+			var doc = XDocument.Load (reader, LoadOptions.SetBaseUri | LoadOptions.SetLineInfo);
+
+			return Parse (doc, collection);
+		}
+
+		public static JavaTypeCollection Parse (string filename, JavaTypeCollection? collection = null)
+		{
+			var doc = XDocument.Load (filename, LoadOptions.SetBaseUri | LoadOptions.SetLineInfo);
+
+			return Parse (doc, collection);
+		}
+
+		static JavaTypeCollection Parse (XDocument doc, JavaTypeCollection? collection = null)
+		{
+			collection ??= new JavaTypeCollection ();
+
+			var root = doc.Root;
+
+			if (root is null)
+				throw new Exception ("Invalid XML file");
+
+			collection.ApiSource = root.XGetAttributeOrNull ("api-source");
+			collection.Platform = root.XGetAttributeOrNull ("platform");
+
+			var packages = new List<JavaPackage> ();
+
+			foreach (var elem in root.Elements ()) {
+				switch (elem.Name.LocalName) {
+					case "package":
+						packages.Add (ParsePackage (elem, collection));
+						break;
+				}
+			}
+
+			// First add all non-nested types
+			foreach (var type in packages.SelectMany (p => p.Types).Where (t => !t.NestedName.Contains ('.')))
+				collection.AddType (type);
+
+			// Add all nested types
+			// This needs to be done ordered from least nested to most nested, in order for nesting to work.
+			// That is, 'android.foo.Blah' needs to be added before 'android.foo.Blah.Bar'.
+			foreach (var type in packages.SelectMany (p => p.Types).Where (t => t.NestedName.Contains ('.')).OrderBy (t => t.FullName.Count (c => c == '.')).ToArray ()) {
+				collection.AddType (type);
+
+				// Remove nested types from Package
+				type.Package.Types.Remove (type);
```

Comment:
This and the associated comment confuse me: why aren't nested types in the package?

Reply:
`Package.Types` isn't "flattened".  It's a "tree", where it contains top-level types, which then contain their nested types.